File: compass-server/src/main/java/org/zalando/compass/domain/persistence/Repository.java
package org.zalando.compass.domain.persistence;

import javax.annotation.Nonnull;
import java.io.IOException;
import java.util.List;
import java.util.Optional;

public interface Repository<T, K, C> {

    boolean create(final T entity) throws IOException;

    default boolean exists(final K id) throws IOException {
        return find(id).isPresent();
    }

    @Nonnull
    default T read(final K id) throws IOException, NotFoundException {
        return find(id).orElseThrow(NotFoundException::new);
    }

    Optional<T> find(final K id) throws IOException;

    List<T> findAll() throws IOException;

    List<T> findAll(final C criteria) throws IOException;

    boolean update(final T entity) throws IOException;

    void delete(final K id) throws IOException;

}

File: src/compass/domain/repository.clj
(ns compass.domain.repository)

(defprotocol DimensionRepository
  (create-dimension [this dimension])
  (read-all-dimensions [this])
  (read-dimensions [this ids])
  (read-dimension [this id])
  ; TODO use a function/patch
  (update-dimension [this dimension])
  (delete-dimension [this dimension]))

(defprotocol KeyRepository
  (create-key [this key])
  (read-all-keys [this])
  (read-keys [this ids])
  (read-key [this id])
  ; TODO use a function/patch
  (update-key [this key])
  (delete-key [this key]))

(defprotocol ValueRepository
  (create-value [this key value])
  (read-all-values [this key])
  ; TODO use a function/patch
  (update-value [this key value])
  (delete-value [this key value]))
